### ISSUE TITLE
Transformers Compatibility

### ISSUE THREAD

**Issue by SpoSer23** [Jyk-122/D-LLM#1]
Hello,
Thank you for this awesome solution in token-level optimization. 

I replicated the code to be transformers compatible in inference and tested the implementation and all is well, The model is Llama 3.2 1.5B.
However, In the case of fine-tuning the model, It always reaches NaN or Inf values in the RMS norm layer in the Router Module.

I checked the data, data flow through the model and tuned the hyperparameters according to the fine-tuning file you used.

Can you tell me what is going wrong during training?

Thanks in advance.

**Comment by Jyk-122** [comment]
Hi

I adapt my training code compatible for transformers(huggingface) locally and give a test on [Llama3.2 1B](https://www.modelscope.cn/models/LLM-Research/Llama-3.2-1B/files)( I didn't find Llama3.2 1.5B ). And mine works fine.

It would be of great help if you tell me more details, e.g. where the NaN values first show up in the network. You may focus on checking whether the attention is calculated normally.

Here's part of my convert code to adapt llms in the form of huggingface compatible for our training code. Hope it's helpful to you.
```
# state dict key mappings from HF's format to meta's format
_FROM_HF_WEIGHT = {
    "model.embed_tokens.weight": "tok_embeddings.weight",
    "model.norm.weight": "norm.weight",
    "lm_head.weight": "output.weight",
    "model.layers.{}.self_attn.q_proj.weight": "layers.{}.attention.wq.weight",
    "model.layers.{}.self_attn.q_proj.bias": "layers.{}.attention.wq.bias",
    "model.layers.{}.self_attn.k_proj.weight": "layers.{}.attention.wk.weight",
    "model.layers.{}.self_attn.k_proj.bias": "layers.{}.attention.wk.bias",
    "model.layers.{}.self_attn.v_proj.weight": "layers.{}.attention.wv.weight",
    "model.layers.{}.self_attn.v_proj.bias": "layers.{}.attention.wv.bias",
    "model.layers.{}.self_attn.o_proj.weight": "layers.{}.attention.wo.weight",
    "model.layers.{}.self_attn.rotary_emb.inv_freq": None,
    "model.layers.{}.mlp.gate_proj.weight": "layers.{}.feed_forward.w1.weight",
    "model.layers.{}.mlp.up_proj.weight": "layers.{}.feed_forward.w3.weight",
    "model.layers.{}.mlp.down_proj.weight": "layers.{}.feed_forward.w2.weight",
    "model.layers.{}.input_layernorm.weight": "layers.{}.attention_norm.weight",
    "model.layers.{}.post_attention_layernorm.weight": "layers.{}.ffn_norm.weight",
}

_FROM_HF_MODELARGS = {
    "hidden_size": "dim",
    "num_hidden_layers": "n_layers",
    "num_attention_heads": "n_heads",
    "num_key_value_heads": "n_kv_heads",
    "vocab_size": "vocab_size",
    "intermediate_size": "intermediate_size",
    "rms_norm_eps": "norm_eps",
    "rope_theta": "rope_theta",
}


def get_mapped_key(key: str, mapping_dict: Dict[str, str]) -> str:
    try:
        if "layers" in key:
            # Replace layer number with "{}" to create key for lookup
            abstract_key = re.sub(r"(\.\d+)", ".{}", key)
            layer_num = re.search(r"\d+", key).group(0)
            new_key = mapping_dict[abstract_key]
            new_key = new_key.format(layer_num)
        else:
            new_key = mapping_dict[key]
    except KeyError as e:
        raise Exception(
            f'Error converting the state dict. Found unexpected key: "{key}". '
            "Please make sure you're loading a checkpoint with the right format. "
        ) from e

    return new_key


def hf_to_meta(
    state_dict: Dict[str, torch.Tensor],
    config: Dict,
    num_heads: int = 32,
    num_kv_heads: int = 32,
    dim: int = 4096,
    head_dim: int = None,
):
    if not isinstance(config, dict):
        config = config.to_dict()

    converted_state_dict = {}

    num_heads = config["num_attention_heads"]
    num_kv_heads = config["num_key_value_heads"]
    dim = config["hidden_size"]

    def _permute(t, n_head):
        return (
            t.reshape(n_head, 2, t.shape[0] // n_head // 2, *t.shape[1:])
            .transpose(1, 2)
            .reshape(*t.shape)
        )

    for key, value in state_dict.items():
        if "rotary_emb.inv_freq" not in key:  # Skip loading the position embeddings
            new_key = get_mapped_key(key, _FROM_HF_WEIGHT)
            if "q_proj" in key:
                value = _permute(value, num_heads)
            elif "k_proj" in key:
                value = _permute(value, num_kv_heads)

            converted_state_dict[new_key] = value

    converted_modelargs = {}
    
    for key, value in config.items():
        if key not in _FROM_HF_MODELARGS.keys():
            continue
        new_key = _FROM_HF_MODELARGS[key]
        converted_modelargs[new_key] = value

    return converted_state_dict, converted_modelargs
```

**Comment by SpoSer23** [comment]
Hi!

It occurs in the RMSNorm layer in the Router Module after Normalization, NaN values start to appear, I will check the attention computation if done correctly.

**Comment by Jyk-122** [comment]
That's weird because NaN tends to show up when divided by 0, which means in RMSNorm the l2-norm of x is 0. Even so, the eps in RMSNorm prevents that. If you think the router goes wrong, try to replace it with simpler network like `nn.Linear(dim, 2)`. If still, try to remove the router or set the output of router == 1, check if lora works fine.

**Comment by SpoSer23** [comment]
I checked the attention computation and it works fine and since we didn't skip yet in inference, they are the same in the base model and the routed model but I tried not using the weight of the RMS norm and it worked fine even produced good output but I still didn't try to use the skipping mechanism in inference yet.

The network is as following:
```
nn.Linear(in_dim, hidden_dim)
RMSNorm()
nn.Linear(hidden_dim, 2)
```

**Comment by Jyk-122** [comment]
do you mean that you changed `return output * self.weight` in RMSNorm to `return output` and the routed model could be trained normally?

**Comment by SpoSer23** [comment]
Yes, Exactly what you mean and produced normal output but I'm still
building the skipping mechanism in token level in inference to work around
it and will check and get back to you.

**Comment by SpoSer23** [comment]
Did you test D-LLM on Llama3.2 1B and Did it make a relative impact on inference time?

**Comment by Jyk-122** [comment]
> Did you test D-LLM on Llama3.2 1B and Did it make a relative impact on inference time?

With given hyper-parameters I trained D-LLM on Llama3.2 1B with Alpaca, the cross-entropy loss on val dataset is about 2.4, and the activated ratio is 0.67 (33% layers on average are pruned). Guess sparsification tends to be harder on smaller models.

**Comment by SpoSer23** [comment]
Hey,

All of the following is stated on my trial with 1B model but I will be
testing the 3B as well:

Without using LoRA tuning, I froze like  98% of the model parameters to
only train the router module across the layers.

But it doesn't show effective impact on smaller models even if it's an
adaptive token mechanism, like with greater max tokens in generation, the
base always had an advantage in first token time generated and total time
generated.

Thanks for the response.

**Comment by Jyk-122** [comment]
> Hey,
> 
> All of the following is stated on my trial with 1B model but I will be
> testing the 3B as well:
> 
> Without using LoRA tuning, I froze like  98% of the model parameters to
> only train the router module across the layers.
> 
> But it doesn't show effective impact on smaller models even if it's an
> adaptive token mechanism, like with greater max tokens in generation, the
> base always had an advantage in first token time generated and total time
> generated.
> 
> Thanks for the response.
> […](#)

LoRA is necessary. Without tuning the parameters of LLM, routers tend to keep all layers because simply removing one layer causes ascended cross-entropy loss. The base model generates faster because the trained d-llm barely skips layers.

**Comment by SpoSer23** [comment]
When you were training, Did the training Cross Entropy loss was so much higher than the validation one with like a 10 to 1 ratio?
Regardless of training hyperparameters

**Comment by Jyk-122** [comment]
> When you were training, Did the training Cross Entropy loss was so much higher than the validation one with like a 10 to 1 ratio? Regardless of training hyperparameters

No such case.

**Comment by SpoSer23** [comment]
I fixed everything in training to be transformers compatible and the model is adjusted to be transformers compatible but I still can't grasp why the normalization layer produces NaNs in the Router Module.

Is it because the difference of dimensions since the router module hidden dimension is dependent on the input dimension of the model in the first linear layer, so it needs normalization in LLMs but not SLMs ?
